processor/transformprocessor/internal/metrics/func_extract_sum_metric.go

```diff
@@ -0,0 +1,119 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package metrics // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/transformprocessor/internal/metrics"
+
+import (
+	"context"
+	"fmt"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlmetric"
+)
+
+type extractSumMetricArguments struct {
+	AggTemp   ottl.Enum `ottlarg:"0"`
+	Monotonic bool      `ottlarg:"1"`
+}
+
+func newExtractSumMetricFactory() ottl.Factory[ottlmetric.TransformContext] {
+	return ottl.NewFactory("extract_sum_metric", &extractSumMetricArguments{}, createExtractSumMetricFunction)
+}
+
+func createExtractSumMetricFunction(_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[ottlmetric.TransformContext], error) {
+	args, ok := oArgs.(*extractSumMetricArguments)
+
+	if !ok {
+		return nil, fmt.Errorf("extractSumMetricFactory args must be of type *extractSumMetricArguments")
+	}
+
+	return extractSumMetric(args.AggTemp, args.Monotonic)
+}
+
+// this interface helps unify the logic for extracting data from different histogram types
+// all supported metric types' datapoints implement it
+type SumCountDataPoint interface {
+	Attributes() pcommon.Map
+	Sum() float64
+	Count() uint64
+	StartTimestamp() pcommon.Timestamp
+	Timestamp() pcommon.Timestamp
+}
+
+func extractSumMetric(aggTempEnum ottl.Enum, monotonic bool) (ottl.ExprFunc[ottlmetric.TransformContext], error) {
```

Comment:
I have some concerns around each of these.

Aggregation temporality: I think we should just copy this from the input metric. Converting aggregation temporality is a stateful operation that should be performed separately, and just setting it without performing the conversion will produce invalid data. I realize this is in some of the other transform processor functions, but it's not clear to me that those will produce valid metrics either.

Monotonicity: We can't make the guarantee that the sum value is monotonic, and the supported input data types have no indicator as to whether they are monotonic. The [OTLP proto](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L415) says that sums are assumed monotonic, but I don't think we can necessarily make that assumption here since the input data may not have been sent over OTLP. Likewise, just setting `monotonic` to false isn't correct either since the input data may very well be monotonic. We will probably just have to leave the option as-is, but I think we should make it more clear that you really need to know your input data to use this function.

Reply:
Both of those make sense to me. The docs already note that you can create invalid metrics by setting this incorrectly - I'll make that more prominent.

Reply:
Ok, so the problem is actually that Summaries do not, in fact, have aggregation temporality as an attribute. So if I want this function to cover all the current types, it needs to stay the way it is. I suppose we could suggest explicitly passing the existing temporality like so `extract_sum_metric(aggregation_temporality, monotonic)`, which will do the right thing in the metric context.

Alternatively, we could use optional parameters once they land, but I'm not sure I want to delay this PR waiting for that. 

WDYT @TylerHelmuth ?

Reply:
I am comfortable with requiring the user to specify whatever aggregation temporality and monotonicity they want.  The transformprocessor is already marked as allowing Unsound Transformations, and the existing functions allow setting these values.

As long as the OTel spec does not have definitions for converting between metric types these functions will always be unsafe.  I would rather give users the freedom to do what they want and give explicit warnings that these functions must only be used when you completely understand the incoming data.

Reply:
>I am comfortable with requiring the user to specify whatever aggregation temporality and monotonicity they want. The transformprocessor is already marked as allowing Unsound Transformations, and the existing functions allow setting these values.

I fully agree with this in principle, but in this case I think these functions would be much more pleasant to use if we provide some guard rails. When using the function with histograms, copying the input aggregation temporality would alleviate most users from having to concern themselves with the temporality of their input data, and would prevent them from unintentionally making a change that would be an uncommon use case. For users who truly want to set the temporality, they can still use `set` to change it.

As for monotonicity, I think users would benefit from a brief explanation of the ramifications of the option. Even something a couple sentences would probably suffice: "the supported input metric types do not support specifying monotonicity, but it is required in the sum metric type, so you should determine whether your input data is monotonic before using this function. This function neither validates nor changes whether input data is monotonic."

>Ok, so the problem is actually that Summaries do not, in fact, have aggregation temporality as an attribute.

That's a good catch.

>So if I want this function to cover all the current types, it needs to stay the way it is. I suppose we could suggest explicitly passing the existing temporality like so `extract_sum_metric(aggregation_temporality, monotonic)`, which will do the right thing in the metric context.
>
>Alternatively, we could use optional parameters once they land, but I'm not sure I want to delay this PR waiting for that.

Could we make the function signature `extract_sum_metric(monotonic, aggregation_temporality)`, then do a non-breaking change later to make `aggregation_temporality` an optional parameter?

Reply:
> Could we make the function signature extract_sum_metric(monotonic, aggregation_temporality), then do a non-breaking change later to make aggregation_temporality an optional parameter?

I'm ok with this.

Reply:
Done. I've also made the warning in the documentation more explicit, as per Evan's suggestion.